osquery: 2.5.2 -> 3.2.2

Ma27 opened the pull request:
###### Motivation for this change

The package was originally broken as reported in #38940 and
facebook/osquery#4257. The latest version (3.x) contains several
important fixes for GCC 7, so now we can compile without a much less
complicated patches.

The following changes were needed to fix the derivation:

* Upgrade `osquery/third-party` to the latest rev to be compliant with
  osquery 3.

* Keep using an override for the AWS SDK (for a lower closure size and
  less compile time), but make the `ec2` API available.

* Added the dependencies `fpm`, `zstd`, `rdkafka`, `rapidjson` to the
  build. `linenoise-ng` is obsolete as it's directly bundled with
  `osquery/third-party`.

* Fixed the linking issue with `gflags` as recommended in the mailing
  list: https://groups.google.com/d/msg/nix-devel/l1blj-mWxtI/J3CwPATBCAAJ

* Dropped the obsolete dependencies `cpp-netlib`, `lz4`, `apt` and
  `devicemapper` (thanks @Infinisil).

* Applied the patch from facebook/osquery#4267 for NixOS support during
  configure phase, override `OSQUERY_PLATFORM` to provide `nixos:version`
  for sandbox and non-NixOS based builds. The `platform-nixos.patch`
  file is now obsolete (thanks @flokli).

* Dropped the newline character from `.version` to avoid `\n` in the
  distro version provided by `osquery`

The patch was rebased against the 3.x branch of `osquery` and contains
mostly old changes. Additionally several testing targets were skipped as
they broke the build.

The functionality has been testing using the following command:

```
mkdir /tmp/osq.log/
./result/bin/osqueryd --pidfile /tmp/osq.pid \
  --database_path /tmp/test.db --logger_path /tmp/osq.log
```

With the daemon running the database can be queried easily using
`./result/bin/osqueryi`.

Fixes ticket #38940
See ticket #36453

Further reference can be gathered from the affected Hydra logs for
the master branch: https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.osquery.x86_64-linux


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39336&attempt_id=3f8d07c3-0b4c-423d-a85d-866d4c52755b)

The following builds were skipped because they don't evaluate on aarch64-linux: osquery

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39336&attempt_id=415b962d-128e-469f-8561-91bf0bcdba74)

Attempted: osquery

<details><summary>Partial log (click to expand)</summary><p>

```
patching script interpreter paths in /nix/store/3n12pzp1znfsr07a7r8a9g9ywwqm160b-aws-sdk-cpp-1.4.33
checking for references to /build in /nix/store/3n12pzp1znfsr07a7r8a9g9ywwqm160b-aws-sdk-cpp-1.4.33...
shrinking RPATHs of ELF executables and libraries in /nix/store/ldc79lsmj10r3imijlddwdi2ks4gj4s7-aws-sdk-cpp-1.4.33-dev
patching script interpreter paths in /nix/store/ldc79lsmj10r3imijlddwdi2ks4gj4s7-aws-sdk-cpp-1.4.33-dev
checking for references to /build in /nix/store/ldc79lsmj10r3imijlddwdi2ks4gj4s7-aws-sdk-cpp-1.4.33-dev...
shrinking RPATHs of ELF executables and libraries in /nix/store/3jcnvd0va0xvh9fkzv433wyj9gfi93bn-aws-sdk-cpp-1.4.33-debug
patching script interpreter paths in /nix/store/3jcnvd0va0xvh9fkzv433wyj9gfi93bn-aws-sdk-cpp-1.4.33-debug
checking for references to /build in /nix/store/3jcnvd0va0xvh9fkzv433wyj9gfi93bn-aws-sdk-cpp-1.4.33-debug...
cannot build derivation '/nix/store/hbjqnqb230fhpc0js447z1ihxbvph4zg-osquery-3.2.2.drv': 1 dependencies couldn't be built
[31;1merror:[0m build of '/nix/store/hbjqnqb230fhpc0js447z1ihxbvph4zg-osquery-3.2.2.drv' failed
```
</p></details>

Ma27 commented:
ok, rebasing onto the latest master seems to have killed it because of newly broken builds...
Will continue to investigate...

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39336&attempt_id=bb65453a-529d-4f04-bee8-cbbbe1a2a790)

The following builds were skipped because they don't evaluate on aarch64-linux: osquery

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39336&attempt_id=b0769b89-afc0-4878-ac58-34368963abdd)

Attempted: osquery

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/rv7s4rcdl660rrv9bf6xyvckq00yvh6a-osquery-3.2.2
shrinking /nix/store/rv7s4rcdl660rrv9bf6xyvckq00yvh6a-osquery-3.2.2/bin/osqueryd
shrinking /nix/store/rv7s4rcdl660rrv9bf6xyvckq00yvh6a-osquery-3.2.2/bin/osqueryi
strip is /nix/store/j75dgadrff2d1fyc4fczmcgqkid2imdx-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/rv7s4rcdl660rrv9bf6xyvckq00yvh6a-osquery-3.2.2/bin
patching script interpreter paths in /nix/store/rv7s4rcdl660rrv9bf6xyvckq00yvh6a-osquery-3.2.2
/nix/store/rv7s4rcdl660rrv9bf6xyvckq00yvh6a-osquery-3.2.2/etc/init.d/osqueryd: interpreter directive changed from "/bin/sh" to "/nix/store/xn5gv3lpfy91yvfy9b0i7klfcxh9xskz-bash-4.4-p19/bin/sh"
checking for references to /build in /nix/store/rv7s4rcdl660rrv9bf6xyvckq00yvh6a-osquery-3.2.2...
/nix/store/rv7s4rcdl660rrv9bf6xyvckq00yvh6a-osquery-3.2.2
```
</p></details>

Infinisil commented:
When I run your command as root for the daemon (and having corrected `--database-path` to `--database_path`) it gives me this error:
```bash
$ sudo result/bin/osqueryd --pidfile /tmp/osq.pid --database_path /tmp/test.db --logger_path /tmp/osq.log
E0422 23:39:28.478080 19220 init.cpp:565] Cannot activate filesystem logger plugin:
  Could not create file: /tmp/osq.log/osqueryd.results.log
```

Even running with root doesn't make it work. 

Here is the strace tail:

    ioprio_set(IOPRIO_WHO_PGRP, 0, IOPRIO_PRIO_VALUE(IOPRIO_CLASS_NONE, 3)) = -1 EINVAL (Invalid argument)
    stat("/etc/osquery/extensions.load", 0x7ffdc8deb0d0) = -1 ENOENT (No such file or directory)
    mmap(NULL, 8392704, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_STACK, -1, 0) = 0x7f3578b0f000
    mprotect(0x7f3578b10000, 8388608, PROT_READ|PROT_WRITE) = 0
    clone(child_stack=0x7f357930b2b0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, parent_tidptr=0x7f357930f9d0, tls=0x7f357930f700, child_tidptr=0x7f357930f9d0) = 29741
    futex(0x7f357930f9d0, FUTEX_WAIT, 29741, NULL[0;31mE0422 23:44:55.228698 29742 init.cpp:565] Cannot activate filesystem logger plugin: Could not create file: /tmp/osq.log/osqueryd.results.log
[m) = 0
    close(4)                                = 0
    exit_group(78)                          = ?
    +++ exited with 78 +++

I thought it might have something to do with me using ZFS, but I also tried it with a tmpfs but I get exactly the same error.

Mic92 commented:
The interesting part where it creates the log file is not included in strace here.

Infinisil commented:
@Mic92 Good point, but even in the full strace it doesn't show up: https://gist.github.com/Infinisil/eb224302b74d3d56a1a4ef2cfce46208 (search for "osq.log"). I think it's the `clone` call that does something in another thread or something (I'm no syscall expert).

Mic92 commented:
Use `-f` for strace so it will follow forks.

Infinisil commented:
Ah yes this makes it a bit clearer: https://gist.github.com/Infinisil/568c8915033266df0d18280951e4e012#file-strace-L6185-L6203

    openat(AT_FDCWD, "/tmp//osqueryd.paul.root.log.ERROR.20180423-005306.2352", O_WRONLY|O_CREAT|O_EXCL, 0664)

It seems to have the logger path set to "" at that point :/

Ma27 commented:
> When I run your command as root for the daemon (and having corrected --database-path to --database_path) it gives me this error:

Ouch, that's been an avoidable typo, thx! :/

> It seems to have the logger path set to "" at that point :/

As stated in the docs (https://github.com/facebook/osquery/blob/3.2.3/docs/wiki/deployment/logging.md) they write files into the `logger_path`, so they expect it to be a directory.

The module creates the `logger_path` in the *preStart* hook of the `systemd` service as well, so I guess that this is the expected behavior.

flokli commented:
@Ma27 Thanks a lot for doing that bump! I'll soon give this a try as well :-)

I recently filed https://github.com/facebook/osquery/pull/4267 adding support for NixOS to `tools/get_platform.py`, mainly for building osquery outside the sandbox (as `/etc/os-release` shouldn't be accessible from the sandbox anyways), but this might help upstreaming some other patches as well, and also get some insight in the "quite odd dependency management" mentioned there - for example, we use `sqlite` from nixpkgs instead of the one from `third-party`.

Can you set `OSQUERY_PLATFORM` instead of applying `pkgs/tools/system/osquery/platform-nixos.patch`?

Infinisil commented:
@Ma27 Ahh, how did I not think of creating that directory, it works indeed. But that error was rather confusing for such a basic problem.

Now that I can verify it working, I'm going to suggest this patch which removes unnecessary dependencies:

```diff
diff --git a/pkgs/tools/system/osquery/default.nix b/pkgs/tools/system/osquery/default.nix
index 8fb2226b7f5..93c971b3998 100644
--- a/pkgs/tools/system/osquery/default.nix
+++ b/pkgs/tools/system/osquery/default.nix
@@ -1,23 +1,14 @@
 { stdenv, lib, fetchFromGitHub, pkgconfig, cmake, pythonPackages
 , udev, audit, aws-sdk-cpp, cryptsetup, lvm2, libgcrypt, libarchive
-, libgpgerror, libuuid, iptables, apt, dpkg, lzma, lz4, bzip2, rpm
+, libgpgerror, libuuid, iptables, dpkg, lzma, bzip2, rpm
 , beecrypt, augeas, libxml2, sleuthkit, yara, lldpd, google-gflags
-, thrift, boost, rocksdb_lite, cpp-netlib, glog, gbenchmark, snappy
-, openssl, file, doxygen, devicemapper
+, thrift, boost, rocksdb_lite, glog, gbenchmark, snappy
+, openssl, file, doxygen
 , gtest, sqlite, fpm, zstd, rdkafka, rapidjson
 }:
 
 let
 
-  aws-sdk-cpp' = aws-sdk-cpp.overrideAttrs (_: {
-    src = fetchFromGitHub {
-      owner = "awslabs";
-      repo = "aws-sdk-cpp";
-      rev = "1.2.7";
-      sha256 = "182wr6j4pqz7g7j9b9cdgkpmflv3b17kl52gkwfz5b2qr8dix2ca";
-    };
-  });
-
   thirdparty = fetchFromGitHub {
     owner = "osquery";
     repo = "third-party";
@@ -44,7 +35,7 @@ stdenv.mkDerivation rec {
   patches = [ ./misc.patch ] ++ lib.optional stdenv.isLinux ./platform-nixos.patch;
 
   nativeBuildInputs = [
-    pkgconfig cmake pythonPackages.python pythonPackages.jinja2
+    pkgconfig cmake pythonPackages.python pythonPackages.jinja2 doxygen fpm
   ];
 
   buildInputs = let
@@ -54,17 +45,16 @@ stdenv.mkDerivation rec {
   in [
     udev audit
 
-    (aws-sdk-cpp'.override {
+    (aws-sdk-cpp.override {
       apis = [ "firehose" "kinesis" "sts" "ec2" ];
       customMemoryManagement = false;
     }).dev
-
-    lvm2 libgcrypt libarchive libgpgerror libuuid iptables.dev apt dpkg
-    lzma lz4 bzip2 rpm beecrypt augeas libxml2 sleuthkit
+    lvm2 libgcrypt libarchive libgpgerror libuuid iptables dpkg
+    lzma bzip2 rpm beecrypt augeas libxml2 sleuthkit
     yara lldpd gflags' thrift boost
-    cpp-netlib glog gbenchmark snappy openssl
-    file doxygen devicemapper cryptsetup
-    gtest sqlite fpm zstd rdkafka rapidjson rocksdb_lite
+    glog gbenchmark snappy openssl
+    file cryptsetup
+    gtest sqlite zstd rdkafka rapidjson rocksdb_lite
   ];
 
   preConfigure = ''
```

- Version 1.2.7 of aws-sdk-cpp isn't needed, the latest one works just fine
- Moves `doxygen` and `fpm` to `nativeBuildInputs`
- Removes dependencies `apt`, `lz4`, `cpp-netlib` and `devicemapper`

Ma27 commented:
> I recently filed facebook/osquery#4267 adding support for NixOS to tools/get_platform.py, mainly for building osquery outside the sandbox (as /etc/os-release shouldn't be accessible from the sandbox anyways), but this might help upstreaming some other patches as well, and also get some insight in the "quite odd dependency management" mentioned there - for example, we use sqlite from nixpkgs instead of the one from third-party.

Awesome! I've just seen the ticket regarding compilation errors on NixOS, but I've obviously missed that. This looks indeed like a much better solution, I'll give it a try!

> Ahh, how did I not think of creating that directory, it works indeed. But that error was rather confusing for such a basic problem.

You're absolutely right! I stumbled about this as well :-)

> Now that I can verify it working, I'm going to suggest this patch which removes unnecessary dependencies:

Thanks! I didn't like the AWS override hack either, so I'll try it out :-)

Ma27 commented:
I fixed the commit with the comments from @Infinisil and @flokli, thanks a lot!

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39336&attempt_id=7ae350db-e2f0-4d1e-9a77-55dbac853fed)

The following builds were skipped because they don't evaluate on aarch64-linux: osquery

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39336&attempt_id=88397ce3-3399-4b91-aabb-627eb659d14b)

Attempted: osquery

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/icbip8m9wsmwh05dgszm5wr0yg4q1j9x-osquery-3.2.2
shrinking /nix/store/icbip8m9wsmwh05dgszm5wr0yg4q1j9x-osquery-3.2.2/bin/osqueryd
shrinking /nix/store/icbip8m9wsmwh05dgszm5wr0yg4q1j9x-osquery-3.2.2/bin/osqueryi
strip is /nix/store/j75dgadrff2d1fyc4fczmcgqkid2imdx-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/icbip8m9wsmwh05dgszm5wr0yg4q1j9x-osquery-3.2.2/bin
patching script interpreter paths in /nix/store/icbip8m9wsmwh05dgszm5wr0yg4q1j9x-osquery-3.2.2
/nix/store/icbip8m9wsmwh05dgszm5wr0yg4q1j9x-osquery-3.2.2/etc/init.d/osqueryd: interpreter directive changed from "/bin/sh" to "/nix/store/xn5gv3lpfy91yvfy9b0i7klfcxh9xskz-bash-4.4-p19/bin/sh"
checking for references to /build in /nix/store/icbip8m9wsmwh05dgszm5wr0yg4q1j9x-osquery-3.2.2...
/nix/store/icbip8m9wsmwh05dgszm5wr0yg4q1j9x-osquery-3.2.2
```
</p></details>

Mic92 commented on pkgs/tools/system/osquery/default.nix:
```diff
@@ -1,59 +1,68 @@
+  patches = [
+    ./misc.patch
+    (fetchpatch {
+      url = "https://patch-diff.githubusercontent.com/raw/facebook/osquery/pull/4267.patch";
```
This url will break if changes are made to the pull request. @flokli could you make a dedicated branch for this that is independent from the pull request and that will not change?

Infinisil commented on .version:
```diff
@@ -1 +1 @@
-18.09
+18.09
```
Looking at the history of this .version file, there were a couple times where this has been updated to remove the trailing newline, and the reason for that is the `nixpkgs/.editorconfig` file which specifies that a trailing newline should be inserted for all files. You can add the following section to that file to prevent further newline insertions for that file:

```ini
[.version]
insert_final_newline = false
```

zimbatm replied:
can this be addressed in a different commit or PR for the git history?

Infinisil commented on pkgs/tools/system/osquery/default.nix:
```diff
@@ -1,59 +1,68 @@
 
   # this is what `osquery --help` will show as the version.
   OSQUERY_BUILD_VERSION = version;
+  OSQUERY_PLATFORM = "nixos;${builtins.readFile "${builtins.toString path}/.version"}";
```
`toString` is available at top-level scope, so you can remove the `builtins.` here.

Ma27 commented on pkgs/tools/system/osquery/default.nix:
```diff
@@ -1,59 +1,68 @@
-    cpp-netlib glog gbenchmark snappy openssl linenoise-ng
-    file doxygen devicemapper cryptsetup
-    gtest sqlite
+    }).dev
```
`.dev` is the output of the derivation which contains the development headers. Just search with `rg "\.dev"` the `nixpkgs` repo and you'll see that this is needed sometimes (Obviously the `aws-sdk` doesn't provide the headers in `out`)

Infinisil replied:
`buildInputs` chooses dev outputs automatically: https://github.com/NixOS/nixpkgs/blob/7c3dc2f53fc837be79426f11c9133f73d15a05c4/pkgs/stdenv/generic/make-derivation.nix#L87

Ma27 commented:
thanks for the suggestions, I'll make the requested changes tonight :-)

infinisil reviewed (approved):
Looks good to me! One last thing though: I would put the .editorconfig update and newline removal in a first commit and the osquery update in a second one.

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39336&attempt_id=4e0303f1-174f-421a-8538-601094a2df69)

Attempted: osquery

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/icbip8m9wsmwh05dgszm5wr0yg4q1j9x-osquery-3.2.2
```
</p></details>

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39336&attempt_id=c66022c9-43b1-4060-a2c6-815c9e4178f3)

The following builds were skipped because they don't evaluate on aarch64-linux: osquery

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

Ma27 commented:
@Mic92 @flokli patch is removed %)

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39336&attempt_id=ad959a5f-841c-4fcf-9d2c-3a20f617c585)

The following builds were skipped because they don't evaluate on aarch64-linux: osquery

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39336&attempt_id=5b1c78bd-164b-42a6-96ab-01f9ea09a990)

Attempted: osquery

<details><summary>Partial log (click to expand)</summary><p>

```
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/3gacg6f39rfpkggbdp86a2mvavkk98l3-osquery-3.2.2
shrinking /nix/store/3gacg6f39rfpkggbdp86a2mvavkk98l3-osquery-3.2.2/bin/osqueryd
shrinking /nix/store/3gacg6f39rfpkggbdp86a2mvavkk98l3-osquery-3.2.2/bin/osqueryi
strip is /nix/store/j75dgadrff2d1fyc4fczmcgqkid2imdx-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/3gacg6f39rfpkggbdp86a2mvavkk98l3-osquery-3.2.2/bin
patching script interpreter paths in /nix/store/3gacg6f39rfpkggbdp86a2mvavkk98l3-osquery-3.2.2
/nix/store/3gacg6f39rfpkggbdp86a2mvavkk98l3-osquery-3.2.2/etc/init.d/osqueryd: interpreter directive changed from "/bin/sh" to "/nix/store/xn5gv3lpfy91yvfy9b0i7klfcxh9xskz-bash-4.4-p19/bin/sh"
checking for references to /build in /nix/store/3gacg6f39rfpkggbdp86a2mvavkk98l3-osquery-3.2.2...
/nix/store/3gacg6f39rfpkggbdp86a2mvavkk98l3-osquery-3.2.2
```
</p></details>

infinisil reviewed (changes requested):
As @zimbatm indicated, the .version thing really needs to be in a different commit or PR

Ma27 committed:
.version: remove final newline

When reading the `nixpkgs` version from `.version` you always have a
`\n` at the end because of the final newline.

This issue exists since b7d15ed and had to be fixed several times
according to the history of `.version`.

Furthermore @Infinisil recommended I explicitly configured
`.editorconfig` to avoid newlines in `.version`.

Ma27 committed:
osquery: 2.5.2 -> 3.2.2

The package was originally broken as reported in #38940 and
facebook/osquery#4257. The latest version (3.x) contains several
important fixes for GCC 7, so now we can compile without a much less
complicated patches.

The following changes were needed to fix the derivation:

* Upgrade `osquery/third-party` to the latest rev to be compliant with
  osquery 3.

* Keep using an override for the AWS SDK (for a lower closure size and
  less compile time), but make the `ec2` API available.

* Added the dependencies `fpm`, `zstd`, `rdkafka`, `rapidjson` to the
  build. `linenoise-ng` is obsolete as it's directly bundled with
  `osquery/third-party`.

* Fixed the linking issue with `gflags` as recommended in the mailing
  list: https://groups.google.com/d/msg/nix-devel/l1blj-mWxtI/J3CwPATBCAAJ

* Dropped the obsolete dependencies `cpp-netlib`, `lz4`, `apt` and
  `devicemapper` (thanks @Infinisil).

* Override `OSQUERY_PLATFORM` to provide `nixos:version`
  for sandbox and non-NixOS based builds. The `platform-nixos.patch`
  file is now obsolete (thanks @flokli).

The patch was rebased against the 3.x branch of `osquery` and contains
mostly old changes. Additionally several testing targets were skipped as
they broke the build.

The functionality has been testing using the following command:

```
mkdir /tmp/osq.log/
./result/bin/osqueryd --pidfile /tmp/osq.pid \
  --database_path /tmp/test.db --logger_path /tmp/osq.log
```

With the daemon running the database can be queried easily using
`./result/bin/osqueryi`.

Fixes ticket #38940
See ticket #36453

Further reference can be gathered from the affected Hydra logs for
the master branch: https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.osquery.x86_64-linux

Ma27 commented:
@Infinisil @zimbatm done %)

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39336&attempt_id=fbe799c2-aa8e-4a23-8a20-021eba9228c8)

Attempted: osquery

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/3gacg6f39rfpkggbdp86a2mvavkk98l3-osquery-3.2.2
```
</p></details>

GrahamcOfBorg commented:
No attempt on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39336&attempt_id=2adc417e-3fe1-4d68-bf2a-b42892643601)

The following builds were skipped because they don't evaluate on aarch64-linux: osquery

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>
